C7/Map/MapViewCamera.cs

```diff
@@ -0,0 +1,116 @@
+using Godot;
+using C7GameData;
+
+namespace C7.Map {
+
+	// MapViewCamera position and zoom should only be modified through the
+	// following methods:
+	// - scaleZoom
+	// - setZoom
+	// - setPosition
+	// This is because these methods will ensure that MapViewCamera handles
+	// world wrapping on the MapView automatically.
+	public partial class MapViewCamera : Camera2D {
+		private readonly float maxZoom = 3.0f;
+		private readonly float minZoom = 0.2f;
+		public float zoomFactor { get; private set; } = 1.0f;
+		private MapView map;
+		private int wrapEdgeTileMargin = 2; // margin in number of tiles to trigger map wrapping
+		private HorizontalWrapState hwrap = HorizontalWrapState.None;
+
+		public void attachToMapView(MapView map) {
+			this.map = map;
+			wrapEdgeMargin = wrapEdgeTileMargin * map.tileSize.X;
+			map.updateAnimations();
+			checkWorldWrap();
+		}
+
+		public override void _Ready() {
+			base._Ready();
+			scaleZoom(zoomFactor);
+		}
+
+		public void scaleZoom(float factor) {
+			zoomFactor = zoomFactor * factor;
+			zoomFactor = Mathf.Clamp(zoomFactor, minZoom, maxZoom);
+			Zoom = Vector2.One * zoomFactor;
+			checkWorldWrap();
+		}
+
+		public void setZoom(float factor) {
+			zoomFactor = Mathf.Clamp(factor, minZoom, maxZoom);
+			Zoom = Vector2.One * zoomFactor;
+			checkWorldWrap();
+		}
+
+		public void setPosition(Vector2 position) {
+			Position = position;
+			checkWorldWrap();
+		}
+		private float wrapEdgeMargin = 0;
+		float wrapRightEdge {get => map.worldEdgeRight - wrapEdgeMargin; }
+		float wrapLeftEdge {get => map.worldEdgeLeft + wrapEdgeMargin; }
+		private bool enteringRightWrap(Rect2 v) => hwrap != HorizontalWrapState.Right && v.End.X >= wrapRightEdge;
+		private bool enteringLeftWrap(Rect2 v) => hwrap != HorizontalWrapState.Left && v.Position.X <= wrapLeftEdge;
+		private bool atEdgeOfRightWrap(Rect2 v) => hwrap == HorizontalWrapState.Right && v.End.X >= wrapRightEdge + map.pixelWidth;
+		private bool atEdgeOfLeftWrap(Rect2 v) => hwrap == HorizontalWrapState.Left && v.Position.X <= wrapLeftEdge - map.pixelWidth;
+		private HorizontalWrapState currentHwrap(Rect2 v) {
+			return v.Position.X <= wrapLeftEdge ? HorizontalWrapState.Left : (v.End.X >= wrapRightEdge ? HorizontalWrapState.Right : HorizontalWrapState.None);
```

Comment:
Would this be better expressed as one of those switch statements or is that possible here? Also is it correct to use v.Position in one case and v.End in the other?

Reply:
not really possible, each predicate is too verbose and it makes the actual logic hard to follow 
position is start https://docs.godotengine.org/en/stable/classes/class_rect2.html#class-rect2-property-end